awscli/customizations/cloudfront.py

```diff
@@ -50,14 +81,103 @@ def __init__(self):
+        return {
+            "CallerReference": unique_string(),
+            "Origins": {"Quantity": 0, "Items": []},
+            "DefaultCacheBehavior": {
+                "TargetOriginId": "placeholder",
+                "ForwardedValues": {
+                    "QueryString": False,
+                    "Cookies": {"Forward": "none"},
+                },
+                "TrustedSigners": {
+                    "Enabled": False,
+                    "Quantity": 0
+                },
+                "ViewerProtocolPolicy": "allow-all",
+                "MinTTL": 0
+            },
+            "Enabled": True,
+            "Comment": "",
+        }
+
+
+class OriginDomainName(ExclusiveArgument):
+    def __init__(self, argument_table):
+        super(OriginDomainName, self).__init__(
+            'origin-domain-name', argument_table,
+            help_text='The domain name for your origin.')
+
+    def add_to_params(self, parameters, value):
+        if value is None:
+            return
+        parameters.setdefault(
+            'DistributionConfig', self.distribution_config_template())
+        origin_id = unique_string(prefix=value)
+        item = {"Id": origin_id, "DomainName": value, "OriginPath": ''}
+        if item['DomainName'].endswith('.s3.amazonaws.com'):
+            # We do not need to detect '.s3[\w-].amazonaws.com' as S3 buckets,
+            # because CloudFront treats GovCloud S3 buckets as custom domain.
+            # http://docs.aws.amazon.com/govcloud-us/latest/UserGuide/setting-up-cloudfront.html
+            item["S3OriginConfig"] = {"OriginAccessIdentity": ""}
+        else:
+            item["CustomOriginConfig"] = {
+                'HTTPPort': 80, 'HTTPSPort': 443,
+                'OriginProtocolPolicy': 'http-only'}
+        parameters['DistributionConfig']['Origins'] = {
+            "Quantity": 1, "Items": [item]}
+        parameters['DistributionConfig']['DefaultCacheBehavior'][
+            'TargetOriginId'] = origin_id
+
+
+class CreateDefaultRootObject(ExclusiveArgument):
+    def __init__(self, argument_table):
+        super(CreateDefaultRootObject, self).__init__(
+            'default-root-object', argument_table, help_text=(
+                'The object that you want CloudFront to return (for example, '
+                'index.html) when a viewer request points to your root URL.'))
+
+    def add_to_params(self, parameters, value):
+        if value is not None:
+            parameters.setdefault(
+                'DistributionConfig', self.distribution_config_template())
+            parameters['DistributionConfig']['DefaultRootObject'] = value
+
+
+class UpdateDefaultRootObject(CreateDefaultRootObject):
+    def __init__(self, context, argument_table):
+        super(UpdateDefaultRootObject, self).__init__(argument_table)
+        self.context = context
+
+    def add_to_params(self, parameters, value):
+        if value is not None:
+            client = self.context['session'].create_client(
```

Comment:
Cool. I think we are fine here. I was worried for a second that the validation of the mutually exclusive arguments happens after ``add_to_params``, meaning we would have an unnecessary client call. But it isn't, so we are fine.